docs/assembly.rst

```diff
@@ -268,3 +278,36 @@ first slot of the array and followed by the array elements.
     Statically-sized memory arrays do not have a length field, but it might be added later
     to allow better convertibility between statically- and dynamically-sized arrays, so
     do not rely on this.
+
+Memory Safety
+=============
+
+Without the use of inline assembly, the compiler can rely on memory to remain in a well-defined
+state at all times. This is especially relevant for :ref:`the new code generation pipeline via Yul IR <ir-breaking-changes>`:
+this code generation path can move local variables from stack to memory to avoid stack-too-deep errors and
+perform additional memory optimizations, if it can rely on certain assumptions about memory use.
+
+While we recommend to always respect Solidity's memory model, it is possible to use memory
+in an incompatible way from inline assembly. Therefore, moving stack variables to memory and additional
+memory optimizations are, by default, disabled in the presence of any inline assembly block.
+
+However, you can specifically annotate an assembly block to indicate that it in fact respects Solidity's memory
+model as follows:
+
+.. code-block:: solidity
+
+    /// @solidity memory-safe-assembly
+    assembly {
+        ...
+    }
+
+In particular, a memory-safe assembly block has to follow the adhere to the following restriction:
+All accessed memory (read or written) was properly allocated. This can either be memory allocated by yourself
+using a mechanism like the ``allocate`` function described above, or memory allocated by Solidity,
+e.g. memory within the bounds of a memory array you reference. The only exception to this is the scratch space
+between memory offset 0 and 64 mentioned above.
```

Comment:
For this we still need to discuss what our actual requirements should really be, so I expect this to still change - for the rest of the docs, a review would be good already, though :-).